semgrep-core/src/engine/Eval_generic.ml

```diff
@@ -137,6 +137,7 @@ let value_of_lit ~code x =
   (* big integers or floats can't be evaluated (Int (None, ...)) *)
   | G.Int (Some i, _t) -> Int i
   | G.Float (Some f, _t) -> Float f
+  | G.Null _ -> Int 0
```

Comment:
? why this? why not add Null to the value?

Reply:
We have this rule in the registry:
```yaml
rules:
  - id: c-string-equality
    patterns:
      - pattern: (char *$X) == (char *$Y)
      - metavariable-comparison:
          metavariable: $X
          comparison: $X != 0
      - metavariable-comparison:
          metavariable: $Y
          comparison: $Y != 0
    message: Using == on char* performs pointer comparison, use strcmp instead
    fix: strcmp($X, $Y) == 0
    languages: [c]
    severity: ERROR
```
and without this it was introducing a regression here:
```c
    // ok:c-string-equality
    if (NULL == s) {
      return 1;
    }
```
Because when `NULL` was not evaluated the `metavariable-comparison` just reduced to `false`, but evaluating `NULL` as `"NULL"` would make `$X != 0` reduce to `true`.

Since `NULL` is actually defined as `(void*)0` I figured out this was an OK solution. Although maybe that should be only when `-lang c`.

If we wanted to restore the exact same behavior here we could, basically comparing `AST _` with anything directly could be reduced to `false`, you would have to use `str()` to do any comparison.

Reply:
I did this instead:
```ocaml
and eval_op op values code =
  match (op, values) with
  | _op, [ AST _; _]
  | _op, [ _; AST _] ->
      (* To compare `AST` values one needs to explicitly use the `str()` function!
       * Otherwise we would introduce regressions. *)
      raise (NotHandled code)
```